File: apps/mobile/src/services/api/tracks.ts
// apps/mobile/src/services/api/tracks.ts
// /tracks API 클라이언트

import { api } from '@services/api'

export interface TrackItem {
  id:            string
  job_id:        string
  song_key:      string
  song_name:     string
  status:        'completed' | 'pending' | 'processing' | 'failed'
  presigned_url: string | null
  created_at:    string    // ISO8601
  completed_at:  string | null
}

export interface TracksListResponse {
  tracks:                     TrackItem[]
  has_pending:                boolean
  completed_since_last_check: boolean
  total:                      number
}

export interface TrackDeleteResponse {
  id:      string
  deleted: boolean
}

export const tracksApi = {
  listTracks: (params?: {
    lastCheckedAt?: string    // ISO8601 (UTC)
    includePresigned?: boolean
  }): Promise<TracksListResponse> => {
    const query: Record<string, string> = {}
    if (params?.lastCheckedAt)
      query.last_checked_at = params.lastCheckedAt
    if (params?.includePresigned !== undefined)
      query.include_presigned = String(params.includePresigned)
    return api.get('/tracks/', { params: query }).then(r => r.data)
  },

  deleteTrack: (trackId: string): Promise<TrackDeleteResponse> =>
    api.delete(`/tracks/${trackId}`).then(r => r.data),
}

File: apps/api/app/schemas/tracks.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel


class TrackItem(BaseModel):
    """홈 화면 트랙 카드 1개 데이터."""

    id: str
    job_id: str
    song_key: str
    song_name: str  # song_key → 한국어 곡명 변환 (서버에서 처리)
    status: str  # completed | pending | processing | failed
    presigned_url: Optional[str] = None  # status='completed'이고 요청 시 포함
    created_at: datetime
    completed_at: Optional[datetime] = None


class TracksListResponse(BaseModel):
    """
    GET /tracks 응답.
    has_pending: 현재 생성 중인 트랙 존재 여부.
    클라이언트가 홈 진입 시 이 플래그를 확인해 폴링 재개 여부 결정.
    last_checked_at: 클라이언트가 이전에 확인한 시각 이후 completed된 트랙이 있으면
    completed_since_last_check=True → "생성 완료 카드" 노출 트리거.
    """

    tracks: List[TrackItem]
    has_pending: bool  # pending 또는 processing 트랙 존재 여부
    completed_since_last_check: bool  # 백그라운드 생성 완료 감지용
    total: int


class TrackDeleteResponse(BaseModel):
    id: str
    deleted: bool
